### PR TITLE
chore(api): Make macros with line breaks span a single line only

### PR DESCRIPTION
**Pull request by bsmth**
### Description

This PR replaces macros that break across multiple lines with single-line invocations.

### Motivation

It's impossible to search through the source for a specific pattern without missing edge cases where macros break across multiple lines. The lack of uniformity of this particular case has resulted in missing certain usages very often, such as (`EmbedLiveSample` calls that contain `.pngs`), i.e.:

```
{{ EmbedLiveSample('Different_shapes_demonstrated', 160, 210,
  "canvas_arc.png") }}
```
It appears that in a lot of cases, this pattern was created through search/replace, and hints at related errors introduced from trying to handle these cases

```
{{macro("thing",
"title")}}`
```

See ``stencil`")}}`` and ```premultipliedAlpha`")}}```

* https://developer.mozilla.org/en-US/docs/Web/API/XRWebGLLayer/framebuffer#how_opaque_framebuffers_are_special
* https://github.com/mdn/content/blob/d77e0bb76025b0a9aed9cff2573588f053fac924/files/en-us/web/api/xrwebgllayer/framebuffer/index.md?plain=1#L55-L60

### CODE REVIEW

File: files/en-us/web/api/audiobuffer/duration/index.md [hunk 1]
```diff
@@ -8,9 +8,8 @@ browser-compat: api.AudioBuffer.duration
 
 {{ APIRef("Web Audio API") }}
 
-The **`duration`** property of the {{ domxref("AudioBuffer")
-    }} interface returns a double representing the duration, in seconds, of the PCM data
-stored in the buffer.
+The **`duration`** property of the {{ domxref("AudioBuffer")}}
```
Comment (OnkarRuikar) [inline files/en-us/web/api/audiobuffer/duration/index.md:11]:
```suggestion
The **`duration`** property of the {{domxref("AudioBuffer")}}
```

File: files/en-us/web/api/audiocontext/createmediastreamsource/index.md [hunk 2]
```diff
@@ -35,11 +34,9 @@ whose media is obtained from the specified source stream.
...
 
-In this example, we grab a media (audio + video) stream from {{
-    domxref("navigator.getUserMedia") }}, feed the media into a {{ htmlelement("video") }}
-element to play then mute the audio, but then also feed the audio into a {{
-    domxref("MediaStreamAudioSourceNode") }}. Next, we feed this source audio into a low
-pass {{ domxref("BiquadFilterNode") }} (which effectively serves as a bass booster),
+In this example, we grab a media (audio + video) stream from {{domxref("navigator.getUserMedia")}}, feed the media into a {{htmlelement("video")}}
+element to play then mute the audio, but then also feed the audio into a {{domxref("MediaStreamAudioSourceNode")}}.
+Next, we feed this source audio into a low pass {{ domxref("BiquadFilterNode") }} (which effectively serves as a bass booster),
 then a {{domxref("AudioDestinationNode") }}.
```
Comment (OnkarRuikar) [inline files/en-us/web/api/audiocontext/createmediastreamsource/index.md:40]:
```suggestion
then a {{domxref("AudioDestinationNode")}}.
```

File: files/en-us/web/api/baseaudiocontext/createbuffer/index.md [hunk 3]
```diff
@@ -10,8 +10,7 @@ browser-compat: api.BaseAudioContext.createBuffer
 
 The `createBuffer()` method of the {{ domxref("BaseAudioContext") }}
 Interface is used to create a new, empty {{ domxref("AudioBuffer") }} object, which
-can then be populated by data, and played via an {{ domxref("AudioBufferSourceNode")
-  }}
+can then be populated by data, and played via an {{ domxref("AudioBufferSourceNode")}}.
```
Comment (OnkarRuikar) [inline files/en-us/web/api/baseaudiocontext/createbuffer/index.md:13]:
```suggestion
can then be populated by data, and played via an {{domxref("AudioBufferSourceNode")}}.
```

File: files/en-us/web/api/baseaudiocontext/createbuffersource/index.md [hunk 4]
```diff
@@ -10,8 +10,8 @@ browser-compat: api.BaseAudioContext.createBufferSource
 
 The `createBufferSource()` method of the {{ domxref("BaseAudioContext") }}
 Interface is used to create a new {{ domxref("AudioBufferSourceNode") }}, which can be
-used to play audio data contained within an {{ domxref("AudioBuffer") }} object. {{
-    domxref("AudioBuffer") }}s are created using
+used to play audio data contained within an {{ domxref("AudioBuffer") }} object.
+{{domxref("AudioBuffer") }}s are created using
```
Comment (OnkarRuikar) [inline files/en-us/web/api/baseaudiocontext/createbuffersource/index.md:14]:
```suggestion
used to play audio data contained within an {{ domxref("AudioBuffer") }} object.
{{domxref("AudioBuffer")}}s are created using
```

File: files/en-us/web/api/biquadfilternode/getfrequencyresponse/index.md [hunk 5]
```diff
@@ -8,8 +8,7 @@ browser-compat: api.BiquadFilterNode.getFrequencyResponse
 
 {{ APIRef("Web Audio API") }}
 
-The `getFrequencyResponse()` method of the {{ domxref("BiquadFilterNode")
-    }} interface takes the current filtering algorithm's settings and calculates the
+The `getFrequencyResponse()` method of the {{ domxref("BiquadFilterNode")}} interface takes the current filtering algorithm's settings and calculates the
```
Comment (OnkarRuikar) [inline files/en-us/web/api/biquadfilternode/getfrequencyresponse/index.md:11]:
```suggestion
The `getFrequencyResponse()` method of the {{domxref("BiquadFilterNode")}} interface takes the current filtering algorithm's settings and calculates the
```

**Review by OnkarRuikar — COMMENTED** [review]
There are more cases of `{{ domxref("abc")}}` or `{{domxref("abc") }}` than I've reportd.

Which one do we prefer?
- `{{ domxref("abc") }}` or
- `{{domxref("abc")}}`

**Comment by hamishwillee** [comment]
> There are more cases of `{{ domxref("abc")}}` or `{{domxref("abc") }}` than I've reportd.
> 
> Which one do we prefer?
> 
> * `{{ domxref("abc") }}` or
> * `{{domxref("abc")}}`

I much prefer the second. It is more compact and the whitespace isn't helpful.

**Comment by OnkarRuikar** [comment]
> > * `{{ domxref("abc") }}` or
> > * `{{domxref("abc")}}`
> 
> I much prefer the second. It is more compact and the whitespace isn't helpful.

We can have this consistent repo wide by adding a custom rule to our linter:
https://github.com/mdn/content/blob/17db3c03142f7077dc335f6f7c127388e2c64442/.markdownlint.jsonc#L175-L179

**Comment by bsmth** [comment]
Thank you for the comments @OnkarRuikar - I agree with all of these suggestions. Shall I handle these in a separate PR? I wanted to keep this one about `\n` newlines only (for reviewers, mainly).

> There are more cases of `{{ domxref("abc")}}` or `{{domxref("abc") }}` than I've reportd.
> 
> Which one do we prefer?
> 
>     * `{{ domxref("abc") }}` or
> 
>     * `{{domxref("abc")}}`

Agree with Hamish on this, +1 to the second format.

**Comment by OnkarRuikar** [comment]
Shall I handle these in a separate PR?

:+1:  Yes, that will be easy to fix and review.

**Comment by wbamberg** [comment]
I'm not exactly against this change, but:
- it's a big PR that's quite hard for humans to review reliably (i.e. without introducing new errors). Do we have any confirmation that this PR doesn't add any macro errors? (and produces the exact same output as before?) 
- it's not exactly true that "It's impossible to search through the source for a specific pattern without missing edge cases where macros break across multiple lines". It's impossible to use grep to look for macro invocations, but that's always going to be unreliable.
- if we are going to rely on macros not containing linebreaks, how do we stop them reappearing?

> It appears that in a lot of cases, this pattern was created through search/replace, and hints at related errors introduced from trying to handle these cases

At least in the two cases you listed, this isn't the case: [the linebreak didn't exist before the error was introduced](https://github.com/mdn/content/commit/bb3a384407f00224e7005fefb64ffd3e20a6953e#diff-481c712318d9da2d6589eb073f0101380c9c12fa9a46c55e394e434c96b46c6f).

Personally I think that rather than using grep it would be better to use tools that are less prone to breaking. For example, I just published https://www.npmjs.com/package/query-kumascript, which uses the KS parser to find macro invocations. It seems like this would be more reliable.

**Comment by OnkarRuikar** [comment]
> it's a big PR that's quite hard for humans to review reliably (i.e. without introducing new errors). Do we have any confirmation that this PR doesn't add any macro errors? (and produces the exact same output as before?)

In a similar case for formal syntax changes I did complete yari builds of before and after changes. And compared the build directories using a visual diff tool Meld. Or we could reduce the batch size to 30 and do 7 PRs.

> if we are going to rely on macros not containing linebreaks, how do we stop them reappearing?

We already have a mechanism in place to lint such custom cases. Following rule in [.markdownlint.jsonc](https://github.com/mdn/content/blob/17db3c03142f7077dc335f6f7c127388e2c64442/.markdownlint.jsonc#L175-L179) should work:
```json
{
"name": "linebreaks-in-macros", 
 "message": "Avoid linebreaks in macro calls", 
 "searchPattern": "/\\{\\{[^{}]*?\n[^{|]*?\\}\\}/g", 
 "searchScope": "text", 
}
```
The regex may need some tuning.

**Comment by hamishwillee** [comment]
FWIW I can cope with varying whitespace in macros, but line breaks within them grate.

**Comment by bsmth** [comment]
Thanks, all. Let me break this one down into smaller PRs. @OnkarRuikar, I feel like 7 is too many PRs, however. 4x55 files with 1LoC changed should be digestible, right?

**Comment by wbamberg** [comment]
So, I'm sorry to harp on about this, but splitting it up doesn't actually reduce the reviewer workload or the chance that errors would be introduced. One way to frame this is: what are the expectations on reviewers for a PR like this? It contains 220 pages: is the reviewer expected to:

- open every page in preview
- find the item generated by the macro
- check it's properly rendered and working, which means for links (which is most of them) means following the link and then adjusting the domain (because preview links don't typically work)

This is a lot of work, and it's very likely that people will make mistakes.

> > it's a big PR that's quite hard for humans to review reliably (i.e. without introducing new errors). Do we have any confirmation that this PR doesn't add any macro errors? (and produces the exact same output as before?)
> 
> In a similar case for formal syntax changes I did complete yari builds of before and after changes. And compared the build directories using a visual diff tool Meld. Or we could reduce the batch size to 30 and do 7 PRs.
> 

Yeah, this is what I had in mind. I think it would be great to smooth the edges for this kind of workflow and make it more common for big cleanup PRs.

> > if we are going to rely on macros not containing linebreaks, how do we stop them reappearing?
> 
> We already have a mechanism in place to lint such custom cases. Following rule in [.markdownlint.jsonc](https://github.com/mdn/content/blob/17db3c03142f7077dc335f6f7c127388e2c64442/.markdownlint.jsonc#L175-L179) should work:
> 
> ```json
> {
> "name": "linebreaks-in-macros", 
>  "message": "Avoid linebreaks in macro calls", 
>  "searchPattern": "/\\{\\{[^{}]*?\n[^{|]*?\\}\\}/g", 
>  "searchScope": "text", 
> }
> ```
> 
> The regex may need some tuning.

I'm not so keen on this. We're going to add more complexity to the build, and more fragile regexes, and why? Why is is so important not to break lines in macro calls? I think the answer here is that people want to use grep+regex to find macro invocations, and I'm questioning whether this is the best approach, and whether having more robust tools (e.g. that are based on the KS parser) would work better for us.

**Comment by bsmth** [comment]
Thanks for sharing the concerns! As far as introducing errors, the flaw system should flag broken things and I don't see problems being introduced by converting multi-line invocations to single-line invocations. I've already broken this up into smaller PRs which I think are fine to review by looking at the diffs alone. There's no rush to merge any of them as long as these get tidied up eventually.

**Comment by hamishwillee** [comment]
FWIW I just reviewed one of these. It was indeed easy to review by inspection. I don't see any problem with converting multi-line invocations to single-line invocations - it is even to my taste. I'd have gone further and removed all arbitrary line breaks.

But we should be pretty clear that while this change isn't risky nor does it really add any concrete value other than to people like me who don't "like" arbitrary line breaks :-). I think if we're going to make these kinds of changes we should put in a process to allow automated output comparision so that minimal human review is required to know that the changes are non-breaking.

Just my two bits.

**Pull request closed without merging**